File: Step-3(Arrays)/Step 3.3 Arrays(Hard)/152_Maximum Product Subarray.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
  int maxProduct(vector<int> &nums)
  {
    int n = nums.size();
    int maxi = INT_MIN;
    int prefixProduct = 1;
    int suffixProduct = 1;
    for (int i = 0; i < n; i++)
    {
      if (prefixProduct == 0)
        prefixProduct = 1;
      if (suffixProduct == 0)
        suffixProduct = 1;

      prefixProduct *= nums[i];
      suffixProduct *= nums[n - i - 1];
      maxi = max(maxi, max(prefixProduct, suffixProduct));
    }
    return maxi;
  }
};

int main()
{
  return 0;
}

File: Step-3(Arrays)/Step 3.3 Arrays(Hard)/152_Maximum Product Subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
  let n = nums.length;
  if (n == 1) return nums[0];
  let maxi = Number.MIN_VALUE;
  let prefixProduct = 1;
  let suffixProduct = 1;
  for (let i = 0; i < n; i++) {
    if (prefixProduct == 0) prefixProduct = 1;
    if (suffixProduct == 0) suffixProduct = 1;

    prefixProduct *= nums[i];
    suffixProduct *= nums[n - i - 1];
    maxi = Math.max(maxi, Math.max(prefixProduct, suffixProduct));
  }
  return maxi;
};
